drivers/led/pca9533.c

```diff
@@ -0,0 +1,410 @@
+}
+
+static int pca9533_led_blink(const struct device *dev, uint32_t led, uint32_t delay_on,
+			     uint32_t delay_off)
+{
+	if (led >= PCA9533_CHANNELS) {
+		LOG_ERR("Invalid LED index: %u", led);
+		return -EINVAL;
+	}
+
+	const struct pca9533_config *config = dev->config;
+	struct pca9533_data *data = dev->data;
+	int ret;
+
+	uint32_t period = delay_on + delay_off;
+
+	if (period < BLINK_MIN_MS || period > BLINK_MAX_MS) {
+		LOG_ERR("Invalid blink period: %u ms (min: %u, max: %u)", period, BLINK_MIN_MS,
+			BLINK_MAX_MS);
+		return -ENOTSUP;
+	}
+
+	/* Calculate duty cycle with overflow protection */
+	uint32_t duty32 = (delay_on * 256U) / period;
+	uint8_t duty = CLAMP(duty32, 0, 255);
+	uint8_t psc = ms_to_psc(period);
+	uint8_t cur = data->led_engine[led];
+
+	/* If LED is sole user of its engine, update in place */
+	if (cur < PCA9533_ENGINES && data->engine_users[cur] == BIT(led)) {
+		/* Only update if parameters changed */
+		if (data->pwm_val[cur] != duty || data->psc_val[cur] != psc) {
+			ret = i2c_reg_write_byte_dt(&config->i2c, cur ? PCA9533_PSC1 : PCA9533_PSC0,
+						    psc);
+			if (ret == 0) {
+				ret = i2c_reg_write_byte_dt(
+					&config->i2c, cur ? PCA9533_PWM1 : PCA9533_PWM0, duty);
+			}
+			if (ret == 0) {
+				data->psc_val[cur] = psc;
+				data->pwm_val[cur] = duty;
+			}
+		}
+		return 0;
+	}
+
+	/* Acquire new engine with desired parameters */
+	uint8_t ch;
+
+	ret = engine_acquire(data, duty, psc, &ch);
+	if (ret) {
+		LOG_WRN("No PWM engine available for LED %u blink", led);
+		return ret;
+	}
+
+	/* If engine is new (no users), program it */
+	if (data->engine_users[ch] == 0) {
+		ret = i2c_reg_write_byte_dt(&config->i2c, ch ? PCA9533_PSC1 : PCA9533_PSC0, psc);
+		if (ret == 0) {
+			ret = i2c_reg_write_byte_dt(&config->i2c, ch ? PCA9533_PWM1 : PCA9533_PWM0,
+						    duty);
+		}
+		if (ret) {
+			LOG_ERR("Failed to program engine %u: %d", ch, ret);
+			return ret;
+		}
+		data->psc_val[ch] = psc;
+		data->pwm_val[ch] = duty;
+	}
+
+	LOG_DBG("LED%u now on engine %u (psc %u duty %u)", led, ch, psc, duty);
+	engine_release(data, led);
+	engine_bind(data, led, ch);
+	return ls_update(&config->i2c, led, ch ? LS_FUNC_PWM1 : LS_FUNC_PWM0);
+}
+
+static int pca9533_led_on(const struct device *dev, uint32_t led)
+{
+	const struct pca9533_config *config = dev->config;
+	struct pca9533_data *data = dev->data;
+
+	if (led >= PCA9533_CHANNELS) {
+		LOG_ERR("Invalid LED index: %u", led);
+		return -EINVAL;
+	}
+
+	int ret = ls_update(&config->i2c, led, LS_FUNC_ON);
+
+	if (ret == 0) {
+		LOG_DBG("LED%u -> ON (steady)", led);
+		engine_release(data, led);
+	}
+
+	return ret;
+}
+
+static int pca9533_led_off(const struct device *dev, uint32_t led)
+{
+	const struct pca9533_config *config = dev->config;
+	struct pca9533_data *data = dev->data;
+
+	if (led >= PCA9533_CHANNELS) {
+		LOG_ERR("Invalid LED index: %u", led);
+		return -EINVAL;
+	}
+
+	int ret = ls_update(&config->i2c, led, LS_FUNC_OFF);
```

Comment:
nit: declare ret with the rest of the local variable at the beginning, that's the preferred way unless you are trying to narrow the variable scope (which you ar enot here), same with the others above